llvm/lib/Transforms/Instrumentation/LowerAllowCheckPass.cpp

```diff
@@ -81,19 +83,32 @@ static bool removeUbsanTraps(Function &F, const BlockFrequencyInfo &BFI,
     return *Rng;
   };
 
-  auto ShouldRemoveHot = [&](const BasicBlock &BB) {
-    return HotPercentileCutoff.getNumOccurrences() && PSI &&
-           PSI->isHotCountNthPercentile(
-               HotPercentileCutoff, BFI.getBlockProfileCount(&BB).value_or(0));
+  auto GetCutoff = [&](const IntrinsicInst *II) {
+    unsigned int cutoff = 0;
+    if (HotPercentileCutoff.getNumOccurrences())
+      cutoff = HotPercentileCutoff;
+    else if (II->getIntrinsicID() == Intrinsic::allow_ubsan_check) {
+      auto *Kind = cast<ConstantInt>(II->getArgOperand(0));
+      if (Kind->getZExtValue() < cutoffs.size())
+        cutoff = cutoffs[Kind->getZExtValue()];
+    }
+
+    return cutoff;
+  };
+
+  auto ShouldRemoveHot = [&](const BasicBlock &BB, unsigned int cutoff) {
+    return PSI && PSI->isHotCountNthPercentile(
+                      cutoff, BFI.getBlockProfileCount(&BB).value_or(0));
```

Comment:
isHotCountNthPercentile does not accept 1000000 which we get from 1.0
I propose to hand this as unconditional remove.

Reply:
Done. I made it fully unconditionally, even if PSI is not available:
```
    return (cutoff == 1000000) ||
           (PSI && PSI->isHotCountNthPercentile(
                       cutoff, BFI.getBlockProfileCount(&BB).value_or(0)));
```